oxenc/common.h

```diff
@@ -32,24 +30,73 @@ concept bt_input_dict_container =
 template <typename Tuple>
 concept tuple_like = requires { std::tuple_size<Tuple>::value; };
 
+namespace detail {
+    template <typename T>
+    inline constexpr bool char_string_type = false;
+    template <basic_char T>
+    inline constexpr bool char_string_type<std::basic_string<T>> = true;
+
+    template <typename T>
+    inline constexpr bool char_view_type = false;
+    template <basic_char T>
+    inline constexpr bool char_view_type<std::basic_string_view<T>> = true;
+}  // namespace detail
+
+template <typename ForwardIt, typename ItValueType = std::iterator_traits<ForwardIt>::value_type>
+concept ordered_pair_iterator =
+        std::forward_iterator<ForwardIt> && std::tuple_size<ItValueType>::value == 2 &&
+        requires(ForwardIt it) {
+            typename ItValueType::first_type;
+            typename ItValueType::second_type;
```

Comment:
This concept isn't working as intended, partly because of this requires expression.

Here's the concept invoked with various types that should, in theory, pass; the ones marked with a `// *` are failing the static_assert:
```C++
static_assert(!ordered_pair_iterator<std::unordered_map<int, int>::iterator>);           // *
static_assert(!ordered_pair_iterator<std::unordered_map<int, int>::const_iterator>);     // *
static_assert(ordered_pair_iterator<std::map<int, int>::iterator>);
static_assert(ordered_pair_iterator<std::map<int, int>::const_iterator>);
static_assert(ordered_pair_iterator<std::vector<std::pair<int, int>>::iterator>);
static_assert(ordered_pair_iterator<std::vector<std::pair<int, int>>::const_iterator>);
static_assert(ordered_pair_iterator<std::vector<std::tuple<int, int>>::iterator>);       // *
static_assert(ordered_pair_iterator<std::vector<std::tuple<int, int>>::const_iterator>); // *
static_assert(ordered_pair_iterator<std::vector<std::array<int, 2>>::iterator>);         // *
static_assert(ordered_pair_iterator<std::vector<std::array<int, 2>>::const_iterator>);   // *
```
The first one is a problem (i.e. we aren't properly failing on unordered_maps).  This is because `::first_type` is a `const Key`, but `unordered_map<const Key, ...>` isn't the same type as`unordered_map<Key, ...>` and so the `is_same<std::unordered_map<...>>` is false for case 1.  That's easily fixed with a `std::remove_const_t`.

The second one is failing because we also want to check for a const_iterator (because if what you have is a `const unordered_map<K, V>&` then your `begin()`/`end()` will be `std::unordered_map<K,V>::const_iterator` rather than `::iterator`.  So this is easily fixed by checking both possible iterator types.

(There's another case I wish we could solve of using a `std::unordered_map<K, V, ...>` where the `...` is non-default: most notably, using a custom hasher, but I don't see any way we can get to that starting from the iterator subtype).

Here's my stab at solving all of these (this makes all the static_asserts pass), split into two concepts because it substantially reduces duplication across the two iterator possibilities:
```c++
template <typename It, typename K, typename V>
concept is_unordered_map_iterator =
        std::same_as<It, typename std::unordered_map<K, V>::iterator> ||
        std::same_as<It, typename std::unordered_map<K, V>::const_iterator>;

template <typename ForwardIt, typename ItValueType = std::iter_value_t<ForwardIt>>
concept ordered_pair_iterator =
        std::forward_iterator<ForwardIt> && std::tuple_size_v<ItValueType> == 2 &&
        !is_unordered_map_iterator<
                ForwardIt,
                std::remove_const_t<std::tuple_element_t<0, ItValueType>>,
                std::tuple_element_t<1, ItValueType>>;
```